lib/Parse/ParseExpr.cpp

```diff
@@ -1045,17 +1045,21 @@ static bool isValidTrailingClosure(bool isExprBasic, Parser &P){
   // the token after the { is on the same line as the {.
   if (P.peekToken().isAtStartOfLine())
     return false;
-  
-  
+
   // Determine if the {} goes with the expression by eating it, and looking
-  // to see if it is immediately followed by '{', 'where', or comma.  If so,
-  // we consider it to be part of the proceeding expression.
+  // to see if it is immediately followed by a token which indicates we should
+  // consider it part of the preceding expression
   Parser::BacktrackingScope backtrack(P);
   P.consumeToken(tok::l_brace);
   P.skipUntil(tok::r_brace);
   SourceLoc endLoc;
   if (!P.consumeIf(tok::r_brace, endLoc) ||
-      P.Tok.isNot(tok::l_brace, tok::kw_where, tok::comma)) {
+      P.Tok.isNot(tok::l_brace, tok::l_square, tok::l_paren, tok::kw_where,
+                  tok::comma, tok::period, tok::period_prefix, tok::kw_is,
+                  tok::kw_as, tok::question_postfix, tok::question_infix,
+                  tok::exclaim_postfix, tok::colon, tok::equal,
+                  tok::oper_postfix, tok::oper_binary_spaced,
+                  tok::oper_binary_unspaced)) {
```

Comment:
Could you insert `isAstStartOfLine()` check for newly added token kinds. For example:
```swift
if foo { something() }
[1, 2, 3].foo()
```
This doesn't look like subscript calling.
```cpp
switch (Tok.getKind()) {
case tok::l_brace:
case tok::kw_where:
case tok::comma:
  return true;
case tok::l_brace:
case tok::l_square:
case ...:
  return !Tok.isAtStartOfLine();
default:
  return false;
}
```